src/str/strs.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id: strs.3,v 1.2 2003/06/17 05:10:55 seth Exp $
.TH STRS 3X "12 June 1993"
.SH NAME
strs_setup, strs_match, strs_done, strs_search - string matching functions
.SH SYNOPSIS
.LP
.nf
.ft B
#include "str.h"
.LP
.ft B
strs_h strs_setup( flags, pattern [, patlen] )
int flags ;
char *pattern ;
.LP
.ft B
char *strs_match( handle, str, len )
strs_h handle ;
char *str ;
int len ;
.LP
.ft B
void strs_done( handle )
strs_h handle ;
.LP
.ft B
char *strs_search( flags, str, len, pattern [, patlen] )
int flags ;
char *str ;
int len ;
char *pattern ;
.SH DESCRIPTION
.LP
These functions perform string matching. They have been designed with
the assumption that one needs to find a certain pattern in a set of
strings. It is also possible to use them to find if a pattern occurs
in a specific string.
.LP
.B strs_setup()
is used to specify the pattern to look for. It returns a
.I handle
which is used in subsequent string matching operations against
the specified
.I pattern.
The
.I flags
argument has two parts: a search method and generic flags.
The available search methods include the following algorithms:
.RS
.TP 15
.SB STRS_BF
brute force algorithm (also called naive in the literature).
.TP
.SB STRS_RK
Rabin-Karp algorithm (probabilistic).
.TP
.SB STRS_KMP
Knuth-Morris-Pratt algorithm.
.TP
.SB STRS_SBM
Simple Boyer-Moore (uses only the last occurrence heuristic).
.TP
.SB STRS_BMH
This is the Boyer-Moore algorithm using the last occurrence heuristic
as modified by Horspool (this is faster than the simple Boyer-Moore).
.TP
.SB STRS_SO
Shift-Or algorithm (this algorithm works only for patterns whose length
does not exceed the number of bits in a word).
.RE
.LP
The default algorithm is the brute force method.
In practice, the fastest algorithm is the
Boyer-Moore-Horspool one.
.LP
The flags that can be specified include:
.RS
.TP 15
.SB STRS_NOMALLOC
do not allocate space for the pattern. This can be specified if
the pattern space will be available during the matching phase
(i.e. do not use this flag if the pattern space was malloc'ed and
you free it before doing any matching).
.TP
.SB STRS_IGNCASE
perform case-insensitive string matching
(the default is case-sensitive matching).
.TP
.SB STRS_NOSWITCH
disallows switching to another search method; a switch to the brute
force algorithm happens if the length of the pattern is less than 4
or if the initialization of the specified search algorithm fails (for
example, when using the shift-or algorithm with a very long pattern).
When this flag is used, no switch happens.
.TP
.SB STRS_PATLEN
is used to explicitly specify the length of the pattern
(with the
.I patlen
argument); normally the pattern is assumed to be NUL-terminated.
.RE
.LP
.B strs_match()
tries to match the string specified by
.I str
against the pattern identified by
.I handle.
Since the length of the string is given by the
.I len
argument the string does not need to be NUL-terminated.
.B strs_done()
should be invoked after all matching against the pattern identified by
.I handle
is done.
.LP
.B strs_search()
is equivalent to:
.LP
.PD .1v
.nf
.RS
h = strs_setup( flags, pattern ) ;
p = strs_match( h, str, len ) ;
strs_done( h ) ;
return( p ) ;
.RE
.PD
.SH "RETURN VALUES"
.LP
.B strs_setup()
returns a search handle on success or
.SM NULL
on failure.
.LP
.B strs_match()
and
.B strs_search()
return a pointer to the first occurrence of the pattern in the string or
.SM NULL
if the pattern does not occur in the string.
.SH "SEE ALSO"
Donald E. Knuth, James H. Morris, Vaughan R. Pratt.
Fast pattern matching in strings.
SIAM Journal on Computing, 6(2):323-350, 1977.
.LP
Richard M. Karp, Michael O. Rabin.
Efficient randomized pattern-matching algorithms.
Technical Report TR-31-81,
Aiken Computation Laboratory, Harvard University, 1981.
.LP
Robert S. Boyer, J. Strother Moore.
A fast string-searching algorithm.
Communications of the ACM, 20(10):762-772, 1977.
.LP
N. Horspool.
Practical fast searching in strings.
Software - Practice and Experience, 10:501-506, 1980.
.LP
R. Baeza-Yates, G.H. Gonnet.
A new approach to text searching.
Proceedings of 12th SIGIR, June 1989.
.LP
Thomas H. Cormen, Charles E. Leiserson, Ronald L. Rivest.
Introduction to Algorithms.
McGraw-Hill 1990.


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/str/strs.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms 
 * and conditions for redistribution.
 */

static char RCSid[] = "$Id: strs.c,v 1.1 2001/05/26 22:04:51 seth Exp $" ;

#include <ctype.h>
#include <stdlib.h>
#include <string.h>

#include "ss_impl.h"
#include "clchack.h"

/*
 * NOTE: The brute force method (with the __strs_bfops) must be always 
 * 		available so that we can switch to it if another method fails.
 */
extern struct ss_ops __strs_bfops ;
extern struct ss_ops __strs_rkops ;
extern struct ss_ops __strs_kmpops ;
extern struct ss_ops __strs_sbmops ;
extern struct ss_ops __strs_bmhops ;
extern struct ss_ops __strs_soops ;

/*
 * NOTE: This table is arranged according to increasing method number.
 *			This allows quick indexing into it using the user-provided
 *			method as a hint:
 *				if ( selection_table[ user_method ].method == user_method )
 *					FOUND
 *				else
 *					DO SEQUENTIAL SEARCH
 *			This allows both quick access and a change of method numbers
 *			in the future without requiring recompilation of programs in
 *			order to work with new versions of the library.
 */
static struct ss_select selection_table[] =
	{
		{ STRS_BF,					&__strs_bfops		},
		{ STRS_RK,					&__strs_rkops		},
		{ STRS_KMP,					&__strs_kmpops		},
		{ STRS_SBM,					&__strs_sbmops		},
		{ STRS_BMH,					&__strs_bmhops		},
		{ STRS_SO,					&__strs_soops		},
		{ 0,							0						}
	} ;

static char identity_map[ ALPHABET_SIZE ] ;
static char upper_to_lower_map[ ALPHABET_SIZE ] ;

static int tables_initialized ;

/*
 * This header is returned when an empty pattern is given to strs_setup.
 * The rest of the functions check ss_patlen and do nothing if that is zero.
 * ss_patlen in this header will be initialized to zero.
 */
static header_s empty_pattern_header ;


PRIVATE void initialize_tables(void)
{
	int i ;

	for ( i = 0 ; i < sizeof( upper_to_lower_map ) ; i++ )
	{
		if ( isascii( i ) && isupper( i ) )
			upper_to_lower_map[ i ] = i + 'a' - 'A' ;
		else
			upper_to_lower_map[ i ] = i ;
		identity_map[ i ] = i ;
	}
}


/*
 * Initializes header
 *
 * Note that 'pattern' does not need to be a NUL-terminated string.
 */
PRIVATE int init(register header_s *hp, int flags, char *pattern, int patlen)
{
	int requested_method = SS_GETMETHOD( flags ) ;
	register struct ss_select *selp ;

	if ( ! tables_initialized )
	{
		initialize_tables() ;
		tables_initialized = TRUE ;
	}

	/*
	 * Initialize header fields
	 */
	SS_FLAGS( hp ) = SS_GETFLAGS( flags ) ;
	SS_PATLEN( hp ) = patlen ;
	if ( SS_SWITCH( hp ) && patlen < 4 )
		SS_OPS( hp ) = &__strs_bfops ;		/* brute force */
	else
	{
		/*
		 * Determine ops
		 */
		if ( selection_table[ requested_method ].sel_method == requested_method )
			selp = &selection_table[ requested_method ] ;
		else
			for ( selp = &selection_table[ 0 ] ; selp->sel_ops ; selp++ )
				if ( requested_method == selp->sel_method )
					break ;
		if ( selp->sel_ops )
			SS_OPS( hp ) = selp->sel_ops ;
		else if ( SS_SWITCH( hp ) )
			SS_OPS( hp ) = &__strs_bfops ;		/* brute force */
		else
			return( SS_ERR ) ;
	}

	if ( SS_MALLOC( hp ) )
	{
		SS_PATTERN( hp ) = malloc( (unsigned)SS_PATLEN( hp ) ) ;
		if ( SS_PATTERN( hp ) == CHAR_NULL )
		{
			(void) free( (char *)hp ) ;
			return( SS_ERR ) ;
		}
		(void) memcpy( SS_PATTERN( hp ), pattern, (int)SS_PATLEN( hp ) ) ;
	}
	else
		SS_PATTERN( hp ) = pattern ;

	/*
	 * If the user asked for case-insensitive search, we create our own
	 * copy of the pattern in lower case. If the pattern is malloc'ed
	 * we overwrite, otherwise we malloc some memory and clear the
	 * STRS_NOMALLOC flag.
	 */
	if ( SS_IGNCASE( hp ) )
	{
		char *new_pattern ;
		register int i ;

		SS_SETMAP( hp, upper_to_lower_map ) ;

		if ( SS_MALLOC( hp ) )
			new_pattern = SS_PATTERN( hp ) ;
		else
		{
			new_pattern = malloc( (unsigned)SS_PATLEN( hp ) + 1 ) ;
			if ( new_pattern == CHAR_NULL )
				return( SS_ERR ) ;
			SS_SETMALLOC( hp ) ;			/* clears the STRS_NOMALLOC flag */
		}
		for ( i = 0 ; i < SS_PATLEN( hp ) ; i++ )
			new_pattern[ i ] = SS_MAP( hp, SS_PATTERN( hp )[ i ] ) ;
		SS_PATTERN( hp ) = new_pattern ;
	}
	else
		SS_SETMAP( hp, identity_map ) ;

	for ( ;; )
	{
		if ( SS_SETUP( hp ) == SS_OK )
			return( SS_OK ) ;
		else
		{
			if ( ! SS_SWITCH( hp ) || SS_OPS( hp ) == &__strs_bfops )
			{
				if ( SS_MALLOC( hp ) )
					(void) free( (char *)hp ) ;
				return( SS_ERR ) ;
			}
			SS_OPS( hp ) = &__strs_bfops ;
		}
	}
}


/*
 * Finalize header
 */
PRIVATE void fini(header_s *hp)
{
	SS_DONE( hp ) ;
	if ( SS_MALLOC( hp ) )
		(void) free( SS_PATTERN( hp ) ) ;
}


/*
 * Create a search handle
 */
strs_h strs_setup( int flags, char *pattern, ... )
{
	header_s		*hp ;
	int			patlen ;
	va_list		ap ;

	hp = HP( malloc( sizeof( *hp ) ) ) ;
	if ( hp == NULL )
		return( NULL_HANDLE ) ;
	
	if ( flags & STRS_PATLEN )
	{
		va_start( ap, pattern ) ;
		patlen = va_arg( ap, int ) ;
		va_end( ap ) ;
	}
	else
		patlen = strlen( pattern ) ;

	if ( patlen == 0 )
		return( (strs_h) &empty_pattern_header ) ;

	if ( init( hp, flags, pattern, patlen ) == SS_OK )
		return( (strs_h)hp ) ;
	else
	{
		free( (char *)hp ) ;
		return( NULL_HANDLE ) ;
	}
}


/*
 * Destroy a search handle
 */
void strs_done(strs_h handle)
{
	header_s *hp = HP( handle ) ;

	if ( SS_PATLEN( hp ) != 0 )
	{
		fini( hp ) ;
		(void) free( (char *) handle ) ;
	}
}


char *strs_match(strs_h handle, char *str, int len)
{
	register header_s *hp = HP( handle ) ;

	if ( SS_PATLEN( hp ) == 0 )
		return( str ) ;
	if ( SS_PATLEN( hp ) > len )
		return( CHAR_NULL ) ;
	return( SS_MATCH( hp, str, len ) ) ;
}



char *strs_search( int flags, char *str, int len, char *pattern, ... )
{
	header_s		t_header ;
	char			*p ;
	int			patlen ;
	va_list		ap ;

	if ( flags & STRS_PATLEN )
	{
		va_start( ap, pattern ) ;
		patlen = va_arg( ap, int ) ;
		va_end( ap ) ;
	}
	else
		patlen = strlen( pattern ) ;

	if ( patlen == 0 )
		return( str ) ;

	if ( patlen > len )
		return( CHAR_NULL ) ;

	if ( init( &t_header, flags | STRS_NOMALLOC, pattern, patlen ) == SS_OK )
	{
		p = SS_MATCH( &t_header, str, len ) ;
		fini( &t_header ) ;
		return( p ) ;
	}
	else
		return( CHAR_NULL ) ;
}


